Minimal viable Play module

jeantil opened the pull request.

jeantil commented:
I am not sure what the project's policy is regarding sample projects. I have created a [repository](https://github.com/jeantil/play-scala-mill-seed) with the mill built play-scala-seed sample project for both 2.6.x and 2.7.x

sorenvalentinjp commented:
I am really looking forward to giving this a spin on my play projects.

lefou commented:
> I am not sure what the project's policy is regarding sample projects. I have created a [repository](https://github.com/jeantil/play-scala-mill-seed) with the mill built play-scala-seed sample project for both 2.6.x and 2.7.x

You could add a link in the module docs page. Also, adding it to the integration sub-project might be a good idea.

jeantil commented:
@sorenvalentinjp you can build mill with this PR by doing : 
```
git clone git@github.com:lihaoyi/mill.git
cd mill 
git fetch origin pull/549/head:play
git co play
ci/publish-local.sh
# Then check the version to include in your import using and picking the latest
ls ~/.ivy2/local/com.lihaoyi/mill-contrib-playlib_2.12/ 
```

This will create a mill-release in your home directory and push the corresponding artifacts in your `.ivy2/local/` repo. 

I would actually like having feedback from someone other than me, on a build that's maybe less trivial than the seed :)

lefou commented on docs/pages/9 - Contrib Modules.md:
```diff
@@ -272,36 +272,240 @@ There's an [example project](https://github.com/lihaoyi/cask/tree/master/example
+Mill commands are targets on a named build. For example if your build is called `core`:
+- compile: `core.compile`
+- run: not implemented. It can be emulated with `mill -w runBackground` but this 
+doesn't do any kind of classloading optimization and may be slower thanPlay's usual `run`. 
```
```suggestion
doesn't do any kind of classloading optimization and may be slower than Play's usual `run`. 
```

lefou commented on contrib/playlib/src/mill/playlib/Version.scala:
```diff
@@ -0,0 +1,17 @@
+private[playlib] trait Version extends Module{
+  def playVersion: T[String]
+
+  private[playlib] def playMinorVersion: T[String] = T {
```
The name suggests, it refers to the single minor version part. Is it possibly a `playBinVersion`? Why do we add the trailing `.0` ?

jeantil replied:
I think minor is correct here, play uses semantic versioning which defines versions as: 
`<major>.<minor>.<patch>` 
This method returns the minor version. I initially added the `.0` because I wasn't sure what I was doing with all the worker key computation for the cross build and bridge worker loading. 
I now realise that the `.0` is useless and can be dropped.

SoerenValentinSilkjaer commented:
I will try to implement this on one of my less trivial builds with scala play and twirl. I will let you know what the result is.

Mogztter commented on contrib/playlib/src/mill/playlib/Layout.scala:
```diff
@@ -0,0 +1,18 @@
+import mill.scalalib._
+
+private[playlib] trait Layout extends JavaModule{
+
```
nitpick, extra newlines

Mogztter commented on docs/pages/9 - Contrib Modules.md:
```diff
@@ -272,36 +272,240 @@ There's an [example project](https://github.com/lihaoyi/cask/tree/master/example
 ### Play Framework
 
-Play framework routes generation support.
+This module adds basic Play framework support to mill: 
```
I think the correct capitalization is "Play Framework"

Mogztter commented on docs/pages/9 - Contrib Modules.md:
```diff
@@ -272,36 +272,240 @@ There's an [example project](https://github.com/lihaoyi/cask/tree/master/example
+- configures mill for Play Framework default directory layout,
+- integrates the Play framework routes compiler,
+- optionally: integrates the Twirl template engine,
+- optionally: configure mill for single module play applications.
```
configures

Mogztter commented on docs/pages/9 - Contrib Modules.md:
```diff
@@ -272,36 +272,240 @@ There's an [example project](https://github.com/lihaoyi/cask/tree/master/example
+ in `mill.playlib`.
+
+The base modules:  
+- `PlayModule` applies the default Play configuration(layout, dependencies, routes compilation, 
```
missing space before the parenthesis

Mogztter commented on docs/pages/9 - Contrib Modules.md:
```diff
@@ -272,36 +272,240 @@ There's an [example project](https://github.com/lihaoyi/cask/tree/master/example
+
+The base modules:  
+- `PlayModule` applies the default Play configuration(layout, dependencies, routes compilation, 
+twirl compilation and akka http server)
```
We should probably capitalize the project names: Twirl compilation and Akka HTTP server

ajrnz commented on contrib/playlib/src/mill/playlib/Version.scala:
```diff
@@ -0,0 +1,17 @@
+    playVersion().split("\\.").take(2).mkString(".")
+  }
+
+  private[playlib] def component(id: String): Target[Dep] = T{
```
This should be a `T.task{}`:

    private[playlib] def component(id: String) = T.task{ ivy"com.typesafe.play::$id::${playVersion()}" }

straight `T{}` blocks can't take parameters. You just just back the first one ever called. I'm not sure how this ever worked for you. It meant that only one of the dependencies was getting pulled in.

ajrnz commented on docs/pages/9 - Contrib Modules.md:
```diff
@@ -272,36 +272,240 @@ There's an [example project](https://github.com/lihaoyi/cask/tree/master/example
+- test: `core.test`
+- dist: not implemented, however you can use the equivalent `core.assembly` 
+command to get a runnable fat jar of the project. The packaging is slightly different though.
+
```
I think you need to be clear there that the server is running in PROD mode and not in DEV. Obvious in hindsight maybe but not when you are trying to get everything working from scratch.

jeantil replied:
Hopefully my new wording is clearer :)

ajrnz commented on docs/pages/9 - Contrib Modules.md:
```diff
@@ -272,36 +272,240 @@ There's an [example project](https://github.com/lihaoyi/cask/tree/master/example
+    │   ├── controllers
+    │   └── views
+    ├── conf
+    ├── logs
```
Maybe put the `routes` file in here so that it is obvious that it still belongs here and is picked up and handled automatically

jeantil replied:
I filtered the tree to only keep the directories but I guess I can be a bit more detailed for the conf part.

jeantil commented:
I realize I must have added the `component` method after I ran all the
tests in a polishing commit.
I have fixed this locally by switching to a task and I have added a unit
test around it to make sure I don't break it again. Will push Monday at the
latest.

Le sam. 2 mars 2019 à 14:34, Andrew Richards <notifications@github.com> a
écrit :

> *@ajrnz* commented on this pull request.
> ------------------------------
>
> In contrib/playlib/src/mill/playlib/Version.scala
> <https://github.com/lihaoyi/mill/pull/549#discussion_r261827049>:
>
> > @@ -0,0 +1,17 @@
> +package mill
> +package playlib
> +
> +import mill.define.Target
> +import mill.scalalib._
> +
> +private[playlib] trait Version extends Module{
> +  def playVersion: T[String]
> +
> +  private[playlib] def playMinorVersion: T[String] = T {
> +    playVersion().split("\\.").take(2).mkString(".")
> +  }
> +
> +  private[playlib] def component(id: String): Target[Dep] = T{
>
> This should be a T.task{}:
>
> private[playlib] def component(id: String) = T.task{ ivy"com.typesafe.play::$id::${playVersion()}" }
>
> straight T{} blocks can't take parameters. You just just back the first
> one ever called. I'm not sure how this ever worked for you. It meant that
> only one of the dependencies was getting pulled in.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/lihaoyi/mill/pull/549#pullrequestreview-209868065>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABZw4P08EgmVYWNUs8IraiZzoSg1Lt-ks5vSn34gaJpZM4a9_FP>
> .
>

ajrnz commented on contrib/playlib/src/mill/playlib/Static.scala:
```diff
@@ -0,0 +1,103 @@
+  }
+
+
+  def start(args: String*) = T.command{ run(args:_*) }
```
I don't think this was meant to be extracted

jeantil replied:
it wasn't this is now fixed, thanks

Mogztter commented on docs/pages/9 - Contrib Modules.md:
````diff
@@ -272,36 +272,252 @@ There's an [example project](https://github.com/lihaoyi/cask/tree/master/example
+        └── controllers
+```
+
+The following compile dependencies will automatically be added to your build:
````
I think we should document the fact that we provide functions to add dependencies.
Maybe with an example.
Eg. if you want to add the `ws` module in your Play application:

```scala
def ivyDeps = super.ivyDeps() ++ Agg(ws())
```

jeantil replied:
I have added a documentation section about this. I considered adding another one about overriding the static assets path but decided against it as it is definitely not something that's routinely done.

Mogztter commented:
@jeantil My application is compiling and running fine, well done :+1: 

A few observations:

When I'm using `core.run` or `core.start` if I kill the process (CTRL + C), the process is still running and the file `RUNNING_PID` is not removed.
Meaning that if I want to run the server again I need to "manually" kill the previous process otherwise I will get the following exceptions:
```
Bind failed because of java.net.BindException"...
```

```
This application is already running (Or delete /path/to/RUNNING_PID file).
```
It might just be a documentation issue and we should state that the process won't be kill.

Also not directly related to this pull request but Play is rightfully throwing the following exception:

```
Oops, cannot start the server.
java.lang.RuntimeException: No application loader is configured. Please configure an application loader either using the play.application.loader configuration property, or by depending on a module that configures one. You can add the Guice support module by adding "libraryDependencies += guice" to your build.sbt.
	at scala.sys.package$.error(package.scala:30)
```

Source: https://github.com/playframework/playframework/blob/524aa8b3803c5b6356d1861acc3b816b741a3370/core/play/src/main/scala/play/api/ApplicationLoader.scala#L53-L55

Not a blocker but we should probably open a pull request in Play to improve the error message otherwise it can be misleading.

Mogztter commented on docs/pages/9 - Contrib Modules.md:
```diff
@@ -272,36 +272,287 @@ There's an [example project](https://github.com/lihaoyi/cask/tree/master/example
+
+Mill commands are targets on a named build. For example if your build is called `core`:
+- compile: `core.compile`
+- run: *NOT Implemented yet*. It can be approximated with `mill -w runBackground` but this 
```
The command should be:

```suggestion
- run: *NOT Implemented yet*. It can be approximated with `mill -w core.runBackground` but this 
```

lihaoyi-databricks commented:
> When I'm using core.run or core.start if I kill the process (CTRL + C), the process is still running and the file RUNNING_PID is not removed.
Meaning that if I want to run the server again I need to "manually" kill the previous process otherwise I will get the following exceptions:

I'm not familiar with this PR, but if you are re-using the logic from `runBackground`, there may be a few tens-of-milliseconds lag before that happens. The expected way to deal with this is to loop on the port for a fraction of the second before giving up so the old process has a chance to notice the change in PID file and kill itself

jeantil commented:
sorry I'm a bit out of time atm, I'll come back to respond to comments asap

lefou commented:
@jeantil Just ping me, if you think that you are done. Looking forward for a merge after your hard work.

jeantil commented:
@lefou There is at least the documentation detail that guillaume raised to commit (but I prefer to rewrite the commit to avoid polluting the history if I can) and I would like to investigate the concerns raised over the runBackground. I had a very busy week this week but I will have literally hours of idle time at the start of next week. 

if the documentation fix is considered enough and then I can quickly fix that tonight or tomorrow morning then you can merge and I open an issue on the runBackground to investigate and see if a fix is needed.

lefou commented:
@jeantil No need to rush. I just don't want to miss the moment you're ready.

Mogztter commented:
> When I'm using core.run or core.start if I kill the process (Ctrl-C), the process is still running and the file RUNNING_PID is not removed.

It's true but as documented I guess:

> starts a server in *PROD* mode which: 
>  - doesn't do any kind of classloading magic (meaning potentially slower restarts)
>  - returns less detailed error messages (no source code extract and line numbers)  

I think we should make it clear that the user will need to kill the process using the `RUNING_PID` file and remove this file to run the server in prod mode again.

I can also reproduce this issue on Linux and macOS with `core.runBackground` but not always :wink: 

```
$ /home/guillaume/mill-release -w core.runBackground
[70/70] core.runBackground 
Watching for changes to 11 files... (Ctrl-C to exit)
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
<Ctrl-C>
```
```
$ /home/guillaume/mill-release -w core.runBackground
[70/70] core.runBackground 
Watching for changes to 11 files... (Ctrl-C to exit)
<Ctrl-C>          
```
```                                                                                            
$ /home/guillaume/mill-release -w core.runBackground
[70/70] core.runBackground 
Watching for changes to 11 files... (Ctrl-C to exit)
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
<Ctrl-C>    
```   
```   
$ /home/guillaume/mill-release -w core.runBackground
[70/70] core.runBackground 
Watching for changes to 11 files... (Ctrl-C to exit)
This application is already running (Or delete /home/guillaume/workspace/constructa/RUNNING_PID file).
```

It seems that if I kill the process during the initialization then the file `RUNNING_PID` is created but the process referenced in this file is gone.

Interestingly if I `Ctrl-C` the command `core.runBackground`, the process won't be killed:

```
$ /home/guillaume/mill-release -w core.runBackground
[70/70] core.runBackground 
Watching for changes to 11 files... (Ctrl-C to exit)
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
SLF4J: Failed to load class "org.slf4j.impl.StaticMDCBinder".
SLF4J: Defaulting to no-operation MDCAdapter implementation.
SLF4J: See http://www.slf4j.org/codes.html#no_static_mdc_binder for further details.
[DEBUG] [03/08/2019 10:54:09.992] [main] [EventStream(akka://reactivemongo)] logger log1-Logging$DefaultLogger started
[DEBUG] [03/08/2019 10:54:09.992] [main] [EventStream(akka://reactivemongo)] Default Loggers started
<Ctrl-C>
```
```
$ cat RUNNING_PID 
4571%                                                                                                                                                                                                                                         
```
```
$ ps -ealf | grep 4571
```

The process is still running:
```
$ netstat -anp | grep 9000 
tcp6       0      0 :::9000                 :::*                    LISTEN      4571/java
```

But if I run `core.runBackground` again it's working and PID changed:

```
$ /home/guillaume/mill-release -w core.runBackground
[70/70] core.runBackground 
Watching for changes to 11 files... (Ctrl-C to exit)
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
SLF4J: Failed to load class "org.slf4j.impl.StaticMDCBinder".
SLF4J: Defaulting to no-operation MDCAdapter implementation.
SLF4J: See http://www.slf4j.org/codes.html#no_static_mdc_binder for further details.
[DEBUG] [03/08/2019 10:59:04.819] [main] [EventStream(akka://reactivemongo)] logger log1-Logging$DefaultLogger started
[DEBUG] [03/08/2019 10:59:04.819] [main] [EventStream(akka://reactivemongo)] Default Loggers started
<Ctrl-C>
```
```
$ cat RUNNING_PID 
4905%
```

And the process is still running:

I believe the prod mode is doing some "sanity" checks:

- If `RUNNING_PID`  exists but the process is gone -> KO `This application is already running...`
- if `RUNNING_PID` does not exist -> OK
- if `RUNNING_PID` exists and process exists -> OK, kill the process and start a new one

Can someone confirm this behavior ?

> I'm not familiar with this PR, but if you are re-using the logic from runBackground, there may be a few tens-of-milliseconds lag before that happens. The expected way to deal with this is to loop on the port for a fraction of the second before giving up so the old process has a chance to notice the change in PID file and kill itself

You might be right...

jeantil commented:
Play will refuse to start if a `RUNNING_PID` file exists it doesn't matter the state of the process whose `PID` is in the RUNNING_PID file.

With the process running :
```bash
> mill -i play-scala-seed.run&
[1] 79781
[73/73] play-scala-seed.run
[info] play.api.Play - Application started (Prod)
[info] p.c.s.AkkaHttpServer - Listening for HTTP on /0:0:0:0:0:0:0:0:9000

> mill -i play-scala-seed.run
[73/73] play-scala-seed.run
This application is already running (Or delete /Users/jean/dev/sdev/src/github/play-scala-seed/RUNNING_PID file).
1 targets failed
play-scala-seed.run subprocess failed
> kill $(cat RUNNING_PID) && ls -l RUNNING_PID
[info] p.c.s.AkkaHttpServer - Stopping server...
-rw-r--r--  1 jean  staff  5  8 mar 14:29 RUNNING_PID
> 1 targets failed
play-scala-seed.run subprocess failed

[1]  + 79781 exit 1     mill -i play-scala-seed.run
> ls -l RUNNING_PID
ls: RUNNING_PID: No such file or directory
```
With the process dead  :
```bash
> mill -i play-scala-seed.run&
[1] 80044
[73/73] play-scala-seed.run
[info] play.api.Play - Application started (Prod)
[info] p.c.s.AkkaHttpServer - Listening for HTTP on /0:0:0:0:0:0:0:0:9000

> kill -9 $(cat RUNNING_PID) && ls -l RUNNING_PID
-rw-r--r--  1 jean  staff  5  8 mar 14:32 RUNNING_PID
> 1 targets failed
play-scala-seed.run subprocess failed

[1]  + 80044 exit 1     mill -i play-scala-seed.run
> ls -l RUNNING_PID
-rw-r--r--  1 jean  staff  5  8 mar 14:32 RUNNING_PID
> ps ax | grep $(cat RUNNING_PID)
> mill -i play-scala-seed.run
[73/73] play-scala-seed.run
This application is already running (Or delete /Users/jean/dev/sdev/src/github/play-scala-seed/RUNNING_PID file).
1 targets failed
play-scala-seed.run subprocess failed
```

If the file is deleted while the process exists you will get a different error. 
````
> mill -i play-scala-seed.run&
[1] 80226
[73/73] play-scala-seed.run
[info] play.api.Play - Application started (Prod)
[info] p.c.s.AkkaHttpServer - Listening for HTTP on /0:0:0:0:0:0:0:0:9000

> rm RUNNING_PID
> mill -i play-scala-seed.run
[73/73] play-scala-seed.run
[info] play.api.Play - Application started (Prod)
Oops, cannot start the server.
akka.stream.impl.io.ConnectionSourceStage$$anon$1$$anon$2: Bind failed because of java.net.BindException: Address already in use
Caused by: java.net.BindException: Address already in use
	at sun.nio.ch.Net.bind0(Native Method)
	at sun.nio.ch.Net.bind(Net.java:433)
	at sun.nio.ch.Net.bind(Net.java:425)
	at sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:223)
	at sun.nio.ch.ServerSocketAdaptor.bind(ServerSocketAdaptor.java:74)
	at akka.io.TcpListener.liftedTree1$1(TcpListener.scala:55)
	at akka.io.TcpListener.<init>(TcpListener.scala:52)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at akka.util.Reflect$.instantiate(Reflect.scala:66)
	at akka.actor.ArgsReflectConstructor.produce(IndirectActorProducer.scala:96)
	at akka.actor.Props.newActor(Props.scala:212)
	at akka.actor.ActorCell.newActor(ActorCell.scala:624)
	at akka.actor.ActorCell.create(ActorCell.scala:650)
	at akka.actor.ActorCell.invokeAll$1(ActorCell.scala:523)
	at akka.actor.ActorCell.systemInvoke(ActorCell.scala:545)
	at akka.dispatch.Mailbox.processAllSystemMessages(Mailbox.scala:283)
	at akka.dispatch.Mailbox.run(Mailbox.scala:224)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:235)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
[error] a.i.TcpListener - Bind failed for TCP channel on endpoint [/0.0.0.0:9000]
java.net.BindException: Address already in use
	at sun.nio.ch.Net.bind0(Native Method)
	at sun.nio.ch.Net.bind(Net.java:433)
	at sun.nio.ch.Net.bind(Net.java:425)
	at sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:223)
	at sun.nio.ch.ServerSocketAdaptor.bind(ServerSocketAdaptor.java:74)
	at akka.io.TcpListener.liftedTree1$1(TcpListener.scala:55)
	at akka.io.TcpListener.<init>(TcpListener.scala:52)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
1 targets failed
play-scala-seed.run subprocess failed
````

Now about mill's behavior when running in client/worker mode (aka without `-i`) more specifically in watch mode.

```bash
mill -w play-scala-seed.runBackground
[74/74] play-scala-seed.runBackground
Watching for changes to 11 files... (Ctrl-C to exit)
[info] play.api.Play - Application started (Prod)
[info] p.c.s.AkkaHttpServer - Listening for HTTP on /0:0:0:0:0:0:0:0:9000
```
in another terminal here is what `pstree -s mill` says : 
```bash 
> pstree -s mill
-+= 00001 root /sbin/launchd
 |-+= 03053 jean /Applications/iTerm.app/Contents/MacOS/iTerm2
 | \-+= 77078 jean /Applications/iTerm.app/Contents/MacOS/iTerm2 --server login -fqp jean
 |   \-+= 77079 root login -fqp jean
 |     \-+= 77080 jean -zsh
 |       \--= 78936 jean java -DMILL_CLASSPATH=/Users/jean/dev/bin/mill -DMILL_VERSION=0.3.6-55-7613e0 -Djna.nosys=true -cp /Users/jean/dev/bin/mill mill.main.client.MillClientMain -w play-scala-seed.runBackground
 \-+- 78837 jean java -DMILL_VERSION=0.3.6-55-7613e0 -cp /Users/jean/dev/bin/mill mill.main.MillServerMain out/mill-worker-1
   \--- 78938 jean java -cp /Users/jean/dev/sdev/src/github/play-scala-seed/conf:/Users/jean/dev/sdev/src/github/play-scala-seed/out/play-scala-seed/webJarResources/dest:/Users/jean/dev/sdev/src/github/play-scala-seed/out/play-scala-seed
```
Now I stop the `mill -w` with `Ctrl-C` (aka SIG_INT), here is what `pstree -s mill` says :
```bash 
> pstree -s mill
-+= 00001 root /sbin/launchd
 \--- 78837 jean java -DMILL_VERSION=0.3.6-55-7613e0 -cp /Users/jean/dev/bin/mill mill.main.MillServerMain out/mill-worker-1
``` 
I restart the `mill -w` background command:
```bash
> mill -w play-scala-seed.runBackground
[74/74] play-scala-seed.runBackground
Watching for changes to 11 files... (Ctrl-C to exit)
[info] play.api.Play - Application started (Prod)
[info] p.c.s.AkkaHttpServer - Listening for HTTP on /0:0:0:0:0:0:0:0:9000
```
pstree says 
```bash
 > pstree -s mill
-+= 00001 root /sbin/launchd
 |-+= 03053 jean /Applications/iTerm.app/Contents/MacOS/iTerm2
 | \-+= 77078 jean /Applications/iTerm.app/Contents/MacOS/iTerm2 --server login -fqp jean
 |   \-+= 77079 root login -fqp jean
 |     \-+= 77080 jean -zsh
 |       \--= 78975 jean java -DMILL_CLASSPATH=/Users/jean/dev/bin/mill -DMILL_VERSION=0.3.6-55-7613e0 -Djna.nosys=true -cp /Users/jean/dev/bin/mill mill.main.client.MillClientMain -w play-scala-seed.runBackground
 \-+- 78837 jean java -DMILL_VERSION=0.3.6-55-7613e0 -cp /Users/jean/dev/bin/mill mill.main.MillServerMain out/mill-worker-1
   \--- 78977 jean java -cp /Users/jean/dev/sdev/src/github/play-scala-seed/conf:/Users/jean/dev/sdev/src/github/play-scala-seed/out/play-scala-seed/webJarResources/dest:/Users/jean/dev/sdev/src/github/play-scala-seed/out/play-scala-seed
```
I interrupt it again with `Ctrl-C` and this is the `pstree -s mill` output 
```bash
> pstree -s mill
-+= 00001 root /sbin/launchd
 \-+- 78837 jean java -DMILL_VERSION=0.3.6-55-7613e0 -cp /Users/jean/dev/bin/mill mill.main.MillServerMain out/mill-worker-1
   \--- 78977 jean java -cp /Users/jean/dev/sdev/src/github/play-scala-seed/conf:/Users/jean/dev/sdev/src/github/play-scala-seed/out/play-scala-seed/webJarResources/dest:/Users/jean/dev/sdev/src/github/play-scala-seed/out/play-scala-seed
```

The second time, the play app was not interrupted when I exited the watched `runBackground` execution. These output are not from a scripted test and I am not that fast a typer so there is easily a couple seconds before I run the pstree again.

Now if I start it again : 
```
> mill -w play-scala-seed.runBackground
[74/74] play-scala-seed.runBackground
Watching for changes to 11 files... (Ctrl-C to exit)
[info] play.api.Play - Application started (Prod)
[info] p.c.s.AkkaHttpServer - Listening for HTTP on /0:0:0:0:0:0:0:0:9000
```
and `pstree -s mill`:
```bash
> pstree -s mill
-+= 00001 root /sbin/launchd
 |-+= 03053 jean /Applications/iTerm.app/Contents/MacOS/iTerm2
 | \-+= 77078 jean /Applications/iTerm.app/Contents/MacOS/iTerm2 --server login -fqp jean
 |   \-+= 77079 root login -fqp jean
 |     \-+= 77080 jean -zsh
 |       \--= 79024 jean java -DMILL_CLASSPATH=/Users/jean/dev/bin/mill -DMILL_VERSION=0.3.6-55-7613e0 -Djna.nosys=true -cp /Users/jean/dev/bin/mill mill.main.client.MillClientMain -w play-scala-seed.runBackground
 \-+- 78837 jean java -DMILL_VERSION=0.3.6-55-7613e0 -cp /Users/jean/dev/bin/mill mill.main.MillServerMain out/mill-worker-1
   \--- 79026 jean java -cp /Users/jean/dev/sdev/src/github/play-scala-seed/conf:/Users/jean/dev/sdev/src/github/play-scala-seed/out/play-scala-seed/webJarResources/dest:/Users/jean/dev/sdev/src/github/play-scala-seed/out/play-scala-seed
```

This suggests that when restarting the runBackground command, mill worker will send a SIG_INT to the previous instance if it is still here and start a new one. 

At least part of the behaviour suggests that the mill client doesn't alway communicate the shutdown order to the mill-worker.

jeantil commented:
@lihaoyi I have not extended run or runBackground at all, I leverage them as is to start a specific main and stop it ...

jeantil commented:
@lefou I think this is it. I have reworded the documentation to warn of the possible RUNNING_PID issues, and clarify that `start` and `run` both start the server in PROD mode.

I don't think the run/runBackground issues raised by @Mogztter actually lie with the PR.

Mogztter commented:
> I don't think the run/runBackground issues raised by @Mogztter actually lie with the PR.

I agree 👍

lihaoyi-databricks commented:
@jeantil Mill only reliably kills the old `runBackground` process when you start a new one, or you `mill clean foo.bar.runBackground` manually. Both cases delete the PID file making the old one self destruct, though there's some 10s-of-milliseconds delay before the old processes notices and self-destructs. Specifically, Mill does not kill the subprocess reliably when you Ctrl-C the Mill client

Could that be what you are seeing?

lefou commented:
Thanks!
